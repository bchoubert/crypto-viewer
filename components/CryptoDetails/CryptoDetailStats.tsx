import React, { memo, FC, useMemo } from 'react';
import { View, Text, StyleSheet } from 'react-native';

import Colors from '../../assets/Colors';

import CryptoViewerIconsMap from '../../assets/fonts/CryptoViewerIconsMap';

import UtilsService from '../../services/Utils.service';

import Stats from '../../models/Stats';
import quoteType from '../../models/QuoteType';
import Crypto from '../../models/Crypto';

const styles = StyleSheet.create({
  stats: {
    flexBasis: 55,
    flexGrow: 0,
    borderTopWidth: 1,
    borderTopColor: Colors.midGray,
    flexDirection: 'column'
  },
  stats_title: {
    flexBasis: 20,
    color: Colors.gray,
    backgroundColor: '#FFFFFF',
    marginHorizontal: 'auto',
    textAlign: 'center',
    marginTop: -11,
    alignSelf: 'center',
    paddingHorizontal: 4
  },
  stats_container: {
    flex: 1,
    flexDirection: 'row',
    marginVertical: 10
  },
  stat: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  stat_number: {
    fontSize: 14
  },
  stat_icon: {
    fontSize: 12,
    paddingRight: 5
  },

  cryptoViewerIcon: {
    fontSize: 20,
    fontFamily: 'crypto-viewer'
  }
});

interface CryptoDetailStatsProps {
  quote: quoteType;
  stats: Stats;
  crypto: Crypto;
}

const CryptoDetailStats: FC<CryptoDetailStatsProps> = ({
  quote,
  stats,
  crypto,
}) => {

  const cryptoColor = useMemo(
    () => UtilsService.getColorFromCrypto(crypto.id),
    [crypto],
  );
  
  if (!stats) {
    return null;
  }
  return (
    <View style={styles.stats}>
      <Text style={styles.stats_title}>24h Stats</Text>
      <View style={styles.stats_container}>

        {/* Highest 24h value section */}
        {(!!stats.high) ?
          (
            <View style={styles.stat}>
              <Text style={{ ...styles.cryptoViewerIcon, ...styles.stat_icon }}>
                {CryptoViewerIconsMap.high.unicode}
              </Text>
              <Text style={{ ...styles.stat_number, color: cryptoColor }}>
                {`${UtilsService.truncateNumber(stats.high, 1)} ${quote.symbol}`}
              </Text>
            </View>
          ) : null
        }

        {/* Lowest 24h value section */}
        {(!!stats.low) ?
          (
            <View style={styles.stat}>
              <Text style={{ ...styles.cryptoViewerIcon, ...styles.stat_icon }}>{CryptoViewerIconsMap.low.unicode}</Text>
              <Text style={{ ...styles.stat_number, color: cryptoColor }}>
                {`${UtilsService.truncateNumber(stats.low, 1)} ${quote.symbol}`}
              </Text>
            </View>
          ) : null
        }

        {/* 24h Volume section */}
        {(!!stats.volume) ?
          (
            <View style={styles.stat}>
              <Text style={{ ...styles.cryptoViewerIcon, ...styles.stat_icon }}>{CryptoViewerIconsMap.volume.unicode}</Text>
              <Text style={{ ...styles.stat_number, color: cryptoColor }}>
                {UtilsService.truncateNumber(stats.volume, 1)}
              </Text>
            </View>
          ) : null
        }
      </View>
    </View>
  );
}

export default memo(CryptoDetailStats);
