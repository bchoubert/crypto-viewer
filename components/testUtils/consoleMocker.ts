/* eslint-disable no-console */
const consoleMocker = {
  consoleError: null,

  errorsToSuppress: [
    'Failed prop type',
  ],

  mockConsole: () => {
    consoleMocker.consoleError = console.error;
    console.error = (message: string) => {
      let printMessage = true;

      consoleMocker.errorsToSuppress.forEach((errorToSuppress: string) => {
        if (message.includes(errorToSuppress)) {
          printMessage = false;
        }
      });

      if (printMessage) {
        consoleMocker.consoleError?.(message);
      }
    };
  },

  restoreConsole: () => {
    if (consoleMocker.consoleError) {
      console.error = consoleMocker.consoleError;
    }
  },
};

export default consoleMocker;
