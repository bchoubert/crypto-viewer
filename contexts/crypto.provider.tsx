import { showError } from "@/services/error.service";
import { fetchCryptoExchangeRates, fetchCryptos } from "@/services/network.service";
import { ICrypto } from "@/types/crypto.types";
import { FC, ReactNode, createContext, memo, useCallback, useContext, useEffect, useMemo, useState } from "react";
import { ToastAndroid } from "react-native";
import { SettingsContext } from "./settings.provider";
import { TranslationsContext } from "./translations.provider";

interface CryptoContextInterface {
  cryptos: ICrypto[];
  isLoading: boolean;
  refreshCryptos: () => void;
}

export const CryptoContext = createContext<CryptoContextInterface>({
  cryptos: [],
  isLoading: false,
  refreshCryptos: () => {},
});

interface CryptoProviderProps {
  children: ReactNode;
}

const CryptoProvider: FC<CryptoProviderProps> = memo(({
  children
}) => {
  const { settings } = useContext(SettingsContext);
  const translation = useContext(TranslationsContext);

  const [isLoading, setLoading] = useState<boolean>(true);
  const [cryptos, setCryptos] = useState<ICrypto[]>([]);

  const refreshCryptos = useCallback(async () => {
    setLoading(true);

    const [fetchedCurrencies, exchangeRatesResponse] = await Promise.all([
      fetchCryptos().catch(showError),
      fetchCryptoExchangeRates(settings.quote).catch(showError),
    ]);

    if (!fetchedCurrencies) {
      ToastAndroid.show(translation.list.error, ToastAndroid.BOTTOM);
      return;
    }
    
    const rates = (exchangeRatesResponse?.data?.rates || {});
    const fetchedCurrenciesWithRates: ICrypto[] = fetchedCurrencies.filter(c => c?.details?.type === 'crypto').map(f => ({
      ...f,
      rate: rates[f.id] ? (1 / parseFloat(rates[f.id])) : undefined,
    }));

    const settingSortAttribute = settings.sortAssets;

    const sortedCurrenciesWithRates = fetchedCurrenciesWithRates.sort((a, b) => {
      const valueA = a[settingSortAttribute].toLowerCase();
      const valueB = b[settingSortAttribute].toLowerCase();

      if (valueA < valueB) {
        return -1;
      }
      if (valueA > valueB) {
        return 1;
      }
      return 0;
    });

    setCryptos(sortedCurrenciesWithRates);
    setLoading(false);
  }, [settings]);

  useEffect(() => { refreshCryptos() }, [settings]);

  const contextValues: CryptoContextInterface = useMemo(() => ({
    cryptos,
    isLoading,
    refreshCryptos,
  }), [cryptos, refreshCryptos, isLoading]);

  return (
    <CryptoContext.Provider value={contextValues}>
      {children}
    </CryptoContext.Provider>
  );
})

export default CryptoProvider;
