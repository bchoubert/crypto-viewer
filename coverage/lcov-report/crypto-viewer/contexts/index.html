
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for crypto-viewer/contexts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> crypto-viewer/contexts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50.63% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>40/79</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/36</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">52.38% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>11/21</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">49.32% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>36/73</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="NavigationProvider.tsx"><a href="NavigationProvider.tsx.html">NavigationProvider.tsx</a></td>
	<td data-value="65.52" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 65%"></div><div class="cover-empty" style="width: 35%"></div></div>
	</td>
	<td data-value="65.52" class="pct medium">65.52%</td>
	<td data-value="29" class="abs medium">19/29</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="62.96" class="pct medium">62.96%</td>
	<td data-value="27" class="abs medium">17/27</td>
	</tr>

<tr>
	<td class="file low" data-value="SettingsProvider.tsx"><a href="SettingsProvider.tsx.html">SettingsProvider.tsx</a></td>
	<td data-value="10.53" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 10%"></div><div class="cover-empty" style="width: 90%"></div></div>
	</td>
	<td data-value="10.53" class="pct low">10.53%</td>
	<td data-value="19" class="abs low">2/19</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="11.76" class="pct low">11.76%</td>
	<td data-value="17" class="abs low">2/17</td>
	</tr>

<tr>
	<td class="file low" data-value="ThemeProvider.tsx"><a href="ThemeProvider.tsx.html">ThemeProvider.tsx</a></td>
	<td data-value="47.83" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 47%"></div><div class="cover-empty" style="width: 53%"></div></div>
	</td>
	<td data-value="47.83" class="pct low">47.83%</td>
	<td data-value="23" class="abs low">11/23</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="23" class="abs low">0/23</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="45.45" class="pct low">45.45%</td>
	<td data-value="22" class="abs low">10/22</td>
	</tr>

<tr>
	<td class="file high" data-value="TranslationProvider.tsx"><a href="TranslationProvider.tsx.html">TranslationProvider.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Mon Sep 06 2021 22:37:01 GMT+0200 (Central European Summer Time)
            </div>
        </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    