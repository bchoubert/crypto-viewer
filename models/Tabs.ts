
export type tabType = 'list' | 'wallet' | 'settings' | 'details';

const Tabs = {
  list: 'list' as tabType,
  details: 'details' as tabType,
  settings: 'settings' as tabType,
  wallet: 'wallet' as tabType,
}

export default Tabs;
