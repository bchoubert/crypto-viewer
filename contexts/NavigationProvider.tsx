import React, {
  createContext, FC, ReactNode, useCallback, useState, useEffect, useMemo, memo, useContext,
} from 'react';
import { StatusBarStyle, BackHandler } from 'react-native';
import { Colors } from 'react-native/Libraries/NewAppScreen';

import Tabs, { TabType } from '../models/Tabs';
import Crypto from '../models/Crypto';
import ColorService from '../services/Color.service';
import { ThemeContext } from './ThemeProvider';

export const NavigationContext = createContext({
  activeTab: Tabs.list as TabType,
  details: null as Crypto | null,
  statusBarColor: Colors.white as string,
  statusBarMode: 'dark-content' as StatusBarStyle,
  changeTab: null as (tabName: TabType, newDetails?: Object) => void | null,
});

interface NavigationProviderProps {
  children: ReactNode;
}

const NavigationProvider: FC<NavigationProviderProps> = ({ children }) => {
  const theme = useContext(ThemeContext);

  const [activeTab, setActiveTab] = useState<TabType>(Tabs.list);

  // Store details about current active tab
  const [details, setDetails] = useState<Crypto | null>(null);

  // StatusBar mode
  const [statusBarColor, setStatusBarColor] = useState<string>(theme.backgroundColor);
  const [statusBarMode, setStatusBarMode] = useState<StatusBarStyle>('dark-content');

  const changeTab = useCallback((tabName: TabType, newDetails: Crypto = null) => {
    setStatusBarColor(tabName === Tabs.details
      ? ColorService.getColorFromCrypto(newDetails.id)
      : theme.backgroundColor);
    setStatusBarMode(tabName === Tabs.details ? 'light-content' as StatusBarStyle : 'dark-content' as StatusBarStyle);
    setDetails(newDetails);
    setActiveTab(tabName);
  }, []);

  const handleBackButton = useCallback(() => {
    if ([Tabs.settings, Tabs.details].includes(activeTab)) {
      changeTab(Tabs.list);
      return true;
    }
    return false;
  }, [activeTab, changeTab]);

  useEffect(() => {
    const backHandler = BackHandler.addEventListener('hardwareBackPress', handleBackButton);
    return () => backHandler.remove();
  }, [activeTab]);

  const contextValues = useMemo(() => ({
    activeTab,
    details,
    statusBarColor,
    statusBarMode,
    changeTab,
  }), [activeTab, details, statusBarColor, statusBarMode, changeTab]);

  useEffect(() => {
    setStatusBarColor(theme.backgroundColor);
  }, [theme]);

  return (
    <NavigationContext.Provider value={contextValues}>
      {children}
    </NavigationContext.Provider>
  );
};

export default memo(NavigationProvider);
