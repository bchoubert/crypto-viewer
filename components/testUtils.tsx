/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';

import NavigationProvider from '../contexts/NavigationProvider';
import { SettingsContext } from '../contexts/SettingsProvider';
import ThemeProvider from '../contexts/ThemeProvider';
import TranslationProvider from '../contexts/TranslationProvider';

import { defaultSettings } from '../models/SettingsType';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));
