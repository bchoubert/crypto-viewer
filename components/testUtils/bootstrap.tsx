/* eslint-disable no-console */
/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';
import * as PassThoughComponents from '../PassThroughComponents';

import NavigationProvider from '../../contexts/NavigationProvider';
import { SettingsContext } from '../../contexts/SettingsProvider';
import ThemeProvider from '../../contexts/ThemeProvider';
import TranslationProvider from '../../contexts/TranslationProvider';

import { defaultSettings } from '../../models/SettingsType';
import consoleMocker from './consoleMocker';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));

beforeAll(() => {
  consoleMocker.mockConsole();
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTimezoneOffset = jest.fn(() => 0);
});

afterAll(() => {
  consoleMocker.restoreConsole();
  jest.resetAllMocks();
});

jest.spyOn(PassThoughComponents, 'Image');
(PassThoughComponents.Image as unknown as jest.Mock).mockImplementation(() => (<div>IMAGE</div>));
