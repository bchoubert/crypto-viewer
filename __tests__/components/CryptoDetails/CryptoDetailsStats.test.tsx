import React from 'react';
import CryptoDetailStats from '../../../components/CryptoDetails/CryptoDetailStats';
import { renderNode } from '../../../components/testUtils/bootstrap';
import { NavigationContext, NavigationContextInterface } from '../../../contexts/NavigationProvider';
import Tabs from '../../../models/Tabs';

describe('<CryptoDetailsStats />', () => {
  const commonProps = {
    buyPrice: 12.345,
    sellPrice: 43.210,
    stats: {
      open: '1',
      high: '1.34',
      low: '2.45',
      last: '3.56',
      volume: '100000',
      volume_30day: '222222',
    },
  };

  const navigationContextProps: NavigationContextInterface = {
    activeTab: Tabs.details,
    statusBarColor: '#334455',
    statusBarMode: 'default',
    changeTab: () => null,
    handleBackAction: () => null,
    details: {
      id: 'btc',
      name: 'Bitcoin',
      price: 555555,
      status: 'online',
      details: {
        type: 'type',
        symbol: 'B',
      },
    },
  };

  it('Prints correctly', () => {
    const { toJSON } = renderNode(
      <NavigationContext.Provider value={navigationContextProps}>
        <CryptoDetailStats
          {...commonProps}
        />
      </NavigationContext.Provider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Basic checks', () => {
    const node = renderNode(
      <NavigationContext.Provider value={navigationContextProps}>
        <CryptoDetailStats
          {...commonProps}
        />
      </NavigationContext.Provider>,
    );

    // Current price
    node.getByText('555555 $');

    // Buy and sell price
    node.getByText('12.35 $');
    node.getByText('43.21 $');

    // Volumes
    node.getByText('100000');

    // Low and high
    node.getByText('2.45 $');
    node.getByText('1.34 $');
  });
});
