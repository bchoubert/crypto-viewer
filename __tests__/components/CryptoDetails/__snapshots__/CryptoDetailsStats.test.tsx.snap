// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoDetailsStats /> Prints correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "marginTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#222222",
        "fontSize": 20,
        "fontWeight": "bold",
        "paddingLeft": 20,
        "paddingRight": 10,
      }
    }
  >
    24h Stats
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginVertical": 10,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#F7931B20",
          "borderRadius": 10,
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": 90,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#222222",
            "flex": 1,
            "fontSize": 28,
          }
        }
      >
        1.34 $
      </Text>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#222222",
                "fontFamily": "crypto-viewer",
                "fontSize": 12,
              }
            }
          >
            
          </Text>
          <Text
            style={
              Object {
                "color": "#222222",
              }
            }
          >
            High
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 10,
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": 90,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "flex": 1,
            "fontSize": 28,
          }
        }
      >
        2.45 $
      </Text>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "crypto-viewer",
                "fontSize": 12,
              }
            }
          >
            
          </Text>
          <Text>
            Low
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#F7931B",
          "borderRadius": 10,
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": 90,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "flex": 1,
            "fontSize": 25,
          }
        }
      >
        100000
      </Text>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontFamily": "crypto-viewer",
                "fontSize": 12,
              }
            }
          >
            
          </Text>
          <Text
            style={
              Object {
                "color": "#FFF",
              }
            }
          >
            Volume
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "#222222",
        "fontSize": 20,
        "fontWeight": "bold",
        "paddingLeft": 20,
        "paddingRight": 10,
      }
    }
  >
    Current Price
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginVertical": 10,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 10,
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": 90,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "flex": 1,
            "fontSize": 28,
          }
        }
      >
        12.35 $
      </Text>
      <Text
        style={
          Object {
            "color": "#333333",
            "fontSize": 14,
          }
        }
      >
        Buy
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#F7931B",
          "borderRadius": 10,
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": 90,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "flex": 1,
            "fontSize": 28,
          }
        }
      >
        555555 $
      </Text>
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
          }
        }
      >
        Price
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#F7931B20",
          "borderRadius": 10,
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": 90,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#222222",
            "flex": 1,
            "fontSize": 28,
          }
        }
      >
        43.21 $
      </Text>
      <Text
        style={
          Object {
            "color": "#222222",
            "fontSize": 14,
          }
        }
      >
        Sell
      </Text>
    </View>
  </View>
</View>
`;
